Normalize city names for easier filtering

vfonic opened the pull request:
The idea is to unify various city spellings so that filter dropdown can work more correctly.

I'm not sure if this is the best approach.
It fixes most of the city names though.

The way it works is:
1. First fix all the city names in the db:
  1.1. Make sure there are no ALLCAPS names
  1.2. Merge different namings or typos for same city (Grad Zagreb, Zaagreb => Zagreb)
2. Add before_validation hook to Ad.city column that tries to find ad with the same (unaccented) city name and use the one from the db as the source of truth

NOTE: We might hit issues with cities like: "Cista", "Čista"

Perhaps easier solution would be to put all the "Osjek" => "Osijek" mappings to `NormalizeColumn` (before_validation)? 🤔
If we do that, we could update all records in the database like this: `Ad.find_each(&:save!)`, since this kicks off the validation before save.

The way I tested this is by pulling all the city names from the live site and then checking what I have in the db locally before and after running the migration.
`Ad.pluck(:city).uniq.sort`


There are more names we could fix, let me know if I should add them one by one:
```
 "Čiovo",  "Čiovo - Trogir"
"Kaštel Lukšić", "Kaštela (Lukšić)",
"Sesvete", "Zagreb Sesvete", "Zagreb,Sesvete",
 "Vukovar (Bogdanovci)", "Vukovar - Bogdanovci",
"Starigrad,Paklenica", "Starigrad-Paklenica", 
```

mmorava commented on app/models/concerns/normalize_column.rb:
```diff
@@ -0,0 +1,25 @@
+    def self.normalize_column(column_name)
+      before_validation -> do
+        value = self[column_name]
+        if value
```
```suggestion
        return unless value
```

vfonic replied:
`return` won't work here as it's not a method, but inside a block. `next unless value` should do.

I'll check. Thanks!

mmorava commented on app/models/concerns/normalize_column.rb:
```diff
@@ -0,0 +1,25 @@
+  included do
+    def self.normalize_column(column_name)
+      before_validation -> do
+        value = self[column_name]
```
```suggestion
        value = self[column_name]&.squish
```

vfonic replied:
I'd leave it as-is. Doesn't change much either way. Thanks!

mmorava commented on db/migrate/20201231092120_normalize_cities.rb:
```diff
@@ -0,0 +1,75 @@
+      Ad.where("unaccent(city) ILIKE unaccent(?)", city).update_all(city: city)
+    end
+
+    puts "City names normalized!"
```
maybe better to use Rails.logger

vfonic replied:
In this case it's more-less the same. On heroku `puts` outputs to stdlog, just like `Rails.logger` does. The only difference would be that with `Rails.logger` you can skip logging by increasing logger level.

vlado commented:
@vfonic Are you on out Slack channel?

vfonic commented:
@vlado no, I'm no Ruby ZG slack, but can't see the channel there

vlado commented:
> @vlado no, I'm no Ruby ZG slack, but can't see the channel there

@markopavlovic Please invite @vfonic to the Slack channel.

vlado commented:
Thanks @vfonic.

I think the ultimate solution would be to have a location select using google maps. We would have everything then and fully normalised. Not sure if this is doable in reasonable timeframe.

This is already helpful but not sure if it is maintainable. Can we pull the list of all cities (including postal codes) from somewhere?

vlado commented:
> @vlado no, I'm no Ruby ZG slack, but can't see the channel there

What is an email you use for Slack?

vfonic commented:
@vlado viktor.fonic@gmail.com

vfonic commented:
@vlado I agree. This is more of a quick fix, which makes the current dropdown, that's already on the live site, more usable and kind-of "working" for now.

There are some gems offering lists of geospacial data. But until we have that implemented, I thought I might "hack" something for the current situation.

berislavbabic commented:
@vfonic @vlado I think I used some version of this list a few years ago, we could export excel into csv with Zip/City https://www.posta.hr/preuzimanje-podataka-o-postanskim-uredima-6543/6543

vfonic commented:
What are we going to do about this PR? Should we close it or merge as a temporary solution until someone builds that CSV Zip/City thing?

This is what I see now on live site. This is not usable:
<img width="1792" alt="Screen Shot 2020-12-31 at 16 49 41" src="https://user-images.githubusercontent.com/67437/103416764-52b28a00-4b88-11eb-8294-9e0854907bef.png">

vlado commented:
@berislavbabic @shime @markopavlovic @fsuste @dejanbj any opinions?

vlado commented:
And we probably need some combobox (dropdown with search).

markopavlovic commented:
I think its fine if we merge this and normalize these cities for now. I think ultimately we want to have a selector (gmaps our another list select), but I vote we have that as a separate issue/task.

ttokic commented:
Hello everyone!
I have written a small Python script which pings https://www.back4app.com/database API and pulls the list of all cities (including postal codes and geoPosition).
You can grab the json from [here](https://www.dropbox.com/s/8p1d4qc42f8cxlb/cities_zip.json?dl=0)

dejanbj commented:
I also think we can go with this for now, but would prefer if we put the existing data normalization logic outside of a db migration - maybe a rake task.

vfonic commented:
Good idea @dejanbj! I'll move the code from the migration to a rake task now.

vlado commented on app/models/concerns/normalize_column.rb:
```diff
@@ -0,0 +1,25 @@
+    def self.normalize_column(column_name)
+      before_validation -> do
+        value = self[column_name]
+        next unless value
```
```suggestion
        next if value.blank?
```

Should save us from checking the database is value is blank.

vfonic replied:
@vlado As-in when it's an empty string?

vlado replied:
Yes, someone could submit the form with empty city or just press space(s) in it (`""`, `" "`, `"       "`). It will not be saved to database and user will see validation error message but we will still do the database call.

If we change it to `next if value.blank?` it will exit early instead.

vlado replied:
And shouldn't it be `return` instead of the `next`?

vfonic replied:
Since it's lambda, it has to be `next`. It's `return` in the last changes.

vlado commented on db/migrate/20201231092120_normalize_cities.rb:
```diff
@@ -0,0 +1,75 @@
+                  ad.city.upcase_first
+                end
+      # skip validation to prevent NormalizeColumn to kick in
+      ad.save!(validate: false)
```
```suggestion
      city = if ad.city.split(/[^[[:word:]]]+/).size == 1
                  ad.city.titleize
                else
                  ad.city.upcase_first
                end
      ad.update_column(:city, city)
```

vfonic commented on app/models/concerns/normalize_column.rb:
```diff
@@ -0,0 +1,25 @@
+
+  included do
+    def self.normalize_column(column_name)
+      before_validation -> do # rubocop:disable Style/Lambda
```
I've opted for this syntax as opposed to:

```ruby
before_validation lambda {
  # ...
}
```

or

```ruby
before_validation(lambda do
  # ....
end)
```

This syntax doesn't work:
```ruby
before_validation lambda do
  # ...
end
```

vlado commented on lib/tasks/ads.rake:
```diff
@@ -0,0 +1,73 @@
+      Veli\ Lošinj
+      Velika\ Gorica
+      Zaprešić Županja
+    ]
```
Apart from moving this to rake task I suggest we also move this to yaml similar to:

```yaml
cities:
  Bjelovar:
    - Bijelovar
    - BELOVAR

  Đakovo:
    - Dakovo
    - Djakovo
```
This way data is separated from the script and we can when needed just update the yaml and rerun the rake task

shime replied:
It would also be cool to have this list in a yaml somewhere and then normalize the city name per that list before saving it.

vfonic replied:
Done. I tested on couple of city names within the YML.

vfonic replied:
Ah, now I see. So these city names query the DB for any case-insensitive, unaccented city name in the db and update that record to match the correct city name.
Should I move all of these to YML and let it do the job?
If I do that, I'll also be able to remove adding the `unaccent` PostgreSQL extension.

vlado reviewed (changes requested):
left few comments

vlado commented:
List of all the cities on production currenty

```
[" Beli manastir", "---orebić", "31551 Belišće", "Andrijaševci ", "Antin", "Antunovac", "BAŠKA VODA", "BIBINJE", "BIOGRAD NA MORU", "BRELA", "BRSEČ", "Bakar", "Bakarac", "Banski Grabovac", "Barban", "Baska", "Baska voda", "Baška", "Baška ", "Baška Voda", "Baška voda", "Bedekovčina", "Bedenica", "Bednja", "Beli Manastir", "Belišće", "Benkovac", "Betina", "Bibinje", "Bihać BiH", "Bijelo Brdo", "Bijelovar", "Biljane Gornje", "Bilje", "Biograd", "Biograd na Moru", "Biograd na Moru,Donje Raštane", "Biograd na moru", "Bisag", "Bjelovar", "Bjelovar ", "Bjelovar,Ribnjacka ", "Bogdanovci kod vukovara", "Bol", "Bosiljevo", "Bosnjaci", "Branjina, Popovac", "Brela", "Brest Pokupski", "Bribir", "Brist", "Brnaze, Sinj", "Brodarica", "Brtonigla ", "Bušetina", "Buševec", "Cakovec", "Cavtat", "Cernik ", "Cestica", "Cres", "Crikvenica", "Crivac ", "Dakovo", "Dalj", "Darda", "Daruvar", "Daruvsr", "Debeljak", "Delnice", "Dingač", "Djakovo", "Donja Bistra", "Donja Zelina", "Donji Andrijevci", "Donji Miholjac", "Draganic", "Dramalj", "Drenovci", "Drenovci, Županja", "Drniš", "Drvenik (kraj Makarske)", "Duboševica Baranja ", "Dubrava", "Dubravice (Skradin)", "Dubrovnik", "Dubrovnik ", "Duga Resa", "Duga Resa ", "Dugi Rat", "Dugi Rat(Omiš)", "Dugo Selo", "Dugo Selo, Rugvica", "Dugo selo, Rugvica", "Fazana", "Fažana Pula", "Feričanci", "Funtana", "GOSPIĆ", "Galovac", "Galovac, Zadar", "Garesnica", "Garešnica ", "Gata", "Gata/Omis", "Glina", "Gorica (Zadar)", "Gornja Hlapa, Dobrinj, Otok Krk", "Gornji Bogićevci", "Gornji Karin", "Gornji Karin(okolica Zadra)", "Gospic", "Gospić", "Gospić ", "Grabovac", "Gracac (skradin)", "Gracac -skradin ", "Grad Zagreb", "Gradec-Vrbovec", "Gračac", "Gračac, Srb", "Gračišće, Istra", "Grubine- Imotski", "Grubišno Polje", "Gundinci", "Gunja", "Gvozd", "Hemuševec", "Herrenberg", "Hrvatski Leskovac", "Hvar", "IVANKOVO", "Igrane", "Imbriovec", "Imotski", "Imotski ", "Imotski , Cista Velika", "Imotski- Runovici ", "Ist", "Ivanec", "Ivanic-Grad", "Ivanić Grad", "Ivanić-Grad", "Ivanićko-Graberje", "Ivankovo", "Ivankovo (vinkovci)", "Ivanska", "JANJINA (PELJEŠAC)", "JESENICE", "Jadranovo", "Jaruge, Slavonski Brod", "Jastrebarsko", "Jelsa", "Jelsa/ Ivan Dolac", "Jesenice", "Jesenice(Omiš) ", "Jezera", "Jezera ", "Josipdol", "Josipovac Punitovački", "KARLOVAC", "KRAVARSKO-NOVO BRDO", "KUTINA", "Kali", "Kanfanar", "Kapelscak", "Karin Gornji", "Karlobag", "Karlovac", "Kastav", "Kastel Luksic", "Kaštel Gomilica", "Kaštel Kambelovac ", "Kaštel Lukšić", "Kaštel Stari", "Kaštel Štafilić", "Kaštel Štafilić ", "Kaštela (Lukšić)", "Kistanje", "Klana", "Klek", "Klis", "Klostar Ivanic", "Kloštar Ivanić", "Knezevi Vinogradi", "Konavle", "Kopacevo", "Kopačevo", "Koprivnica", "Koprivnica ", "Koprivnica - Drnje", "Koprivnički Bregi", "Korenica", "Korivnica", "Korčula", "Korčula - Brna", "Košutarica ,44324 Jasenovac", "Kraljevec na Sutli", "Kraljevica", "Kranj", "Krapina", "Krapinske Toplice", "Krašić", "Krivi Put ", "Križ", "Križevci", "Križevci ", "Krk", "Krk, Punat", "Kršan", "Kukljica", "Kupljenovo (uz Zaprešić)", "Kutina", "LASLOVO/OSIJEK", "Ladimirevci", "Ladimirevci, Valpovo", "Lasinja", "Lasinja ", "Legrad", "Lekenik", "Linardići, Krk", "Lipa", "Lipik", "Lipovac", "Lisac,Ston", "Ljubešćica", "Lokvičići ", "Lovinac", "Lovran", "Ludbreg", "Lukač", "Lučko", "Lužani", "MAKARSKA", "MATULJI", "MOSTAR", "MOŠĆENIČKA Draga ", "MURTER", "Makarska", "Makarska ", "Mala Ludina", "Mala Subotica", "Malinska", "Mandre", "Mandre, otok Pag", "Marcana(Pula) ", "Marija Bistrica", "Marija Gorica", "Marina", "Marina-Trogir", "Markovac Našički", "Martinšćica", "Marusevec", "Marčana Kavran", "Maslenica", "Matulji", "Matulji / Opatija", "Matulji-Rukavac", "Medulin", "Medulin ", "Medulin/Loborika", "Metković", "Međugorje", "Mijesto Brckovljani", "Milna! Otok Brac", "Mimice", "Misenicka draga ", "Molunat", "Mostar", "Mošćenička Draga", "Mraclin", "Mrzović", "Mursko Sredisce", "Mursko Središče", "Murter ", "N. Zagreb", "Nard", "Našice-Lađanska", "Nedelišće", "Neorić-Sutina (Split) ", "Neum", "Nin", "Nin, Žerava", "Njivice", "Nova Gradiska", "Nova Gradiška", "Nova Gradiška ", "Nova Kapela", "Novalja", "Novi Vinodolski", "Novi Vinodolski ", "Novi Vinodolski - Bribir", "Novigrad", "Novigrad (Zadarski)", "Novigrad Podravski", "Novigrad Podravski ", "Novigrad, Zadar", "Novigrad/Paljuv", "Novo Virje", "Novo Čiče", "Novoselec", "Novoselec ", "Novska", "Novska ", "OSIJEK", "OSIJEK ", "Obedišće", "Obrovac", "Obrovac ", "Ogulin", "Ogulin ", "Oklaj", "Okolica Sinja ", "Okolica Splita", "Okrug Gornji", "Okrug Gornji, Ciovo", "Okučani", "Omis", "Omiš", "Omiš,Podašpilje", "Opatija", "Opatija ", "Oprtalj", "Orahovica", "Orebic", "Orehovica", "Oroslavje", "Osijek", "Osijek ", "Osijek(Kopačevo)", "Osjek", "Otok Hvar", "Otok Krk", "Otok Pasman kod Zadra", "Otok Pasman kraj Zadra", "Otok Rab", "Otok Vir", "Otok vir ", "Otočac", "Otočac ", "Ozalj", "PAKOSTANE", "PETRINJA", "PIROVAC", "PITOMAČA", "PREMANTURA ", "PRIMOŠTEN", "Pag", "Pakostane", "Pakoštane", "Pakoštane ", "Pakrac", "Pakrac ", "Palešnik", "Palešnik (Garešnica) ", "Palešnik - Garešnica", "Petrijanec", "Petrinja", "Petrovci kraj Vukovara", "Pirovac", "Pisak", "Pisarovina", "Pitomača", "Pleternica", "Pleternica ", "Plitvice", "Plitvicka jezera", "Plitvička Jezera", "Ploce", "Ploče", "Pobri", "Podaca", "Podaca - Makarska rivijera ", "Podgora", "Podgora ", "Podstrana", "Podstrana , Split", "Podstrana Split", "Podstrana(kraj Splita)", "Polača", "Poličnik", "Poljica-Marina", "Popovac", "Popovača", "Porec", "Poreč", "Poreč ", "Posedarje", "Posedarje ", "Posedarje - Islam Latinski", "Posedarje Slivnica", "Povlja otok Brac", "Povljana", "Povljana(Pag)", "Požega", "Požega ", "Pregrada", "Pregrada ", "Preko,poljana", "Prelog", "Prezid", "Prezid, Gorski kotar", "Primošten", "Primošten ", "Privlaka", "Promajna ", "Pula", "Pula ", "Puska", "Rab", "Radenci", "Rajevo Selo", "Rajić", "Rakitovica", "Rakovica", "Račišće 109", "Raša", "Ražanj-Rogoznica", "Retkovci", "Rešetari", "Rijeka", "Rijeka ", "Rogoznica", "Rogoznica ", "Rogoznica /Šibenik ", "Rovinj", "Rovinj j(Rovigno)", "Rukavac-Matulji", "SAVUDRIJA", "SPLITSKA OTOK BRAČ", "Saborsko", "Samobor", "Seget Vranjica", "Seget Vranjica ", "Selce", "Seline,Starigrad-Paklenica", "Selo Jezero Posavsko", "Senj", "Sesvete", "Sesvete ", "Severin na Kupi", "Sibenik", "Sinj", "Sirač ", "Sisak", "Sisak (Caprag) ", "Siče", "Skopje ", "Skradin", "Skradin ", "Slatina", "Slatina ", "Slatina, Gornji Miholjac", "Slavonski Brod", "Slavonski Brod ", "Slavonski Brod (Donja Bebrina)", "Slavonski Brod/Gornja Vrba", "Slavonski Kobaš ", "Slavonski brod (Podcrkavlje)", "Slovenija, adlešiči", "Slunj", "Slunj ", "Solin", "Soline", "Soljani, Vukovarsko-Srijemska", "Spisic Bukovica (Virovitica) ", "Spl", "Split", "Split ", "Split/Grude", "Split/Split", "Sreser", "Srijane", "Srima ", "St. Toplice", "Stara Gradiska", "Stara Gradiška ", "Stara Novalja", "Stari Jankovci", "Stari Jankovci/Vinkovci", "Stari Perkovci", "Starigrad Dol", "Starigrad Paklenica", "Starigrad,Paklenica", "Starigrad-Paklenica", "Staro Petrovo Selo", "Staro Petrovo Selo, Ljudevita Posavskog 1", "Staro petrovo selo", "Stobrec", "Strizivojna", "Stupnik", "Suhopolje ", "Sukošan", "Sukošan (Zadar)", "Sukošan-Zadar", "Sumartin", "Supetar", "Supetar, Brac", "Sutivan, Brač", "Sveti Filip i Jakov", "Sveti Ivan Zelina, Beloslavec", "Sveti Martin na Muri", "Sveti Petar na Moru", "Sveti Petar na moru", "Sveti križ začretje", "Sveti martin na muri", "Svibje", "Svirče Hvar", "TROGIR", "Terezino Polje", "Tisno", "Tkon", "Tomislavgrad", "Topusko", "Tordinci", "Tribalj", "Tribunj", "Trilj", "Trogir", "Trosnj", "Trpanj", "Trpanj Pelješac", "Tršće,Čabar Gorski Kotar", "Tucepi", "Tuheljske Toplice", "Tučepi", "Udbina", "Umag", "Umag ", "VRSI", "Vabriga", "Valpovo ", "Varadin", "Varazdin", "Varazdin ", "Varazdinu ", "Varaždin", "Varaždin ", "Varaždin, Radovec", "Varaždinske Toplice", "Varaždinske Toplice ", "Vela Luka", "Veli Losinj", "Velika", "Velika Gorica", "Velika Gorica ", "Velika Gorica (Kobilić) ", "Velika Kopanica", "Velika Kopanica ", "Velika gorica ", "Veliki Grđevac / Bjelovarsko-bilogorska", "Veliko Trojstvo", "Verunić", "Vinišće", "Vinjerac ", "Vinkovci", "Vinkovci ", "Vinkovci (Sopot) ", "Vinogradci", "Vir", "Vir ", "Virje", "Virovitica", "Visoko ", "Vivodina, Ozalj", "Viškovo", "Vocin ", "Vodice", "Vodice ", "Vodice Srima", "Vodice, ČISTA VELIKA", "Vodnjan", "Vodnjan(Dignano)", "Vodnjan~Peroj", "Vozilici", "Vođinci", "Vođinci ", "Vranjic", "Vratišinec-Međimurje", "Vrbanja(Županja)", "Vrbanja,Hrvatska", "Vrbnik", "Vrbnik ", "Vrbovec", "Vrgorac", "Vrlika, Garjak", "Vrpolje ", "Vrsar ", "Vrsi", "Vukovar", "Vukovar (Bogdanovci)", "Vukovar - Bogdanovci", "Vukovar, Ilača", "Waterford", "ZADAR", "ZADAR ", "ZAGREB", "ZAGREB( SESVETE)", "Zaagreb", "Zabok", "Zadar", "Zadar ", "Zadar (Briševo)", "Zadar (Crna Punta) ", "Zadar Preko- Poljana", "Zadar( Sukošan )", "Zadar, Sv.Filip i Jakov", "Zadarska zupanija", "Zagreb", "Zagreb ", "Zagreb - Dubrava", "Zagreb - Vrbovec", "Zagreb / Lucko", "Zagreb Kašina", "Zagreb Perjavica", "Zagreb Sesvete", "Zagreb,Sesvete", "Zagrebacka zupanija Varazdinska Krapinsko zagorska ", "Zaprešić", "Zašrešić", "Zelina", "Zemunik Donji (Zadar)", "Zlatar Bistrica", "Zubovići ", "bregana", "dalj", "dubrovnik", "gg", "komin neretva", "mali Lošinj", "mali losinj", "marčana", "okučani", "primosten", "seget", "sibenik", "soljani", "split", "tuzla", "zadar", "zagreb", "Čakovec", "Čavle", "Čavle ", "Čavle-Cernik 13(okolica Rijeke)", "Čazma", "Čačinci", "Čačinci ", "Čađavački Lug / Slatina", "Čepin", "Čepinski Martinci", "Čilipi", "Čiovo", "Čiovo - Trogir", "Čiovo, Zagreb", "Črnkovci ", "Črnkovec", "čakovec", "Đakovo", "Đelekovec ", "Đurđevac", "Đurđevac ", "ŠIBENIK", "ŠIBENik", "ŠMRIKA", "Šandrovac", "Šarengrad", "Šegotići, Marčana, Pula ", "Šestanovac", "Šibenik", "Šibenik ", "Šibenik, Rogoznica", "Šilo (Krk)", "Širitovci/Drniš", "Šišinec", "Škabrnja", "Škrip (otok Brač)", "Škropeti", "Šmrika", "Šolta", "Štitar", "Štrigova", "šibenik", "Ždrelac (Otok Pašman)", "Živogošće", "Županja"]
```

vfonic commented:
There's another thing missing which could be added: we could add an index to `city` column, as that's what we're using to query the database here:
https://github.com/vlado/earthquake-croatia/pull/41/files#diff-6963b9ed355c11d5af18e4f0ad28a208c038e5fd456554174100743455af412aR20

vfonic commented:
Ok, pushed the fixes. This should include everything now.

vfonic commented on app/models/concerns/normalize_city_name.rb:
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+module NormalizeCityName
```
I had to rename this to be less generic `NormalizeColumn` => `NormalizeCityName` because on line 29 below, we use `cities_mapping.yml` to fetch incorrect to correct city names mapping.

vfonic commented on app/models/concerns/normalize_city_name.rb:
```diff
@@ -0,0 +1,41 @@
+                    city.upcase_first
+                  end
+
+      normalized_from_yml = normalize_from_yml
+      return if normalized_from_yml
```
This looks a bit ugly. Any suggestions how to write this in a better way?

vlado replied:
`normalize_from_yml or normalize_from_other_db_records`

vlado commented:
Hi @vfonic can we make sure this is mergable so we do cleanup of existing cities ASAP? We can finish polishing and improving this in a second step then.

vfonic commented:
@vlado I'll make this mergeable today (Sunday). Can you please run `Ad.pluck(:city).uniq` on live site again if possible? Thank you!

vlado commented:
> @vlado I'll make this mergeable today (Sunday). Can you please run `Ad.pluck(:city).uniq` on live site again if possible? Thank you!

```
[" Beli manastir", "---orebić", "31551 Belišće", "Andrijaševci ", "Antin", "Antunovac", "BAŠKA VODA", "BIBINJE", "BIOGRAD NA MORU", "BRELA", "BRSEČ", "BUJE", "Bakar", "Bakarac", "Bale", "Banski Grabovac", "Barban", "Baska", "Baska voda", "Batina", "Baška", "Baška ", "Baška Voda", "Baška voda", "Bedekovčina", "Bedenica", "Bednja", "Beli Manastir", "Belišće", "Benkovac", "Betina", "Bibinje", "Bihać BiH", "Bijelo Brdo", "Bijelovar", "Biljane Gornje", "Bilje", "Biograd", "Biograd na Moru", "Biograd na Moru,Donje Raštane", "Biograd na moru", "Bisag", "Bjelovar", "Bjelovar ", "Bjelovar,Ribnjacka ", "Bobovac", "Bogdanovci kod vukovara", "Bol", "Bosiljevo", "Bosnjaci", "Branjina, Popovac", "Brela", "Brest Pokupski", "Bribir", "Brist", "Brnaze, Sinj", "Brodarica", "Brtonigla ", "Budva", "Bušetina", "Buševec", "Cakovec", "Cavtat", "Cernik ", "Cestica", "Cres", "Crikvenica", "Crivac ", "Dakovo", "Dalj", "Darda", "Daruvar", "Daruvsr", "Debeljak", "Delnice", "Dingač", "Djakovo", "Donja Bistra", "Donja Lomnica", "Donja Zelina", "Donji Andrijevci", "Donji Miholjac", "Draganic", "Dramalj", "Drenovci", "Drenovci, Županja", "Drniš", "Drvenik (kraj Makarske)", "Duboševica Baranja ", "Dubrava", "Dubravice (Skradin)", "Dubrovnik", "Dubrovnik ", "Duga Resa", "Duga Resa ", "Dugi Rat", "Dugi Rat(Omiš)", "Dugo Selo", "Dugo Selo, Rugvica", "Dugo selo", "Dugo selo, Rugvica", "Dugopolje", "Fazana", "Fažana Pula", "Feričanci", "Funtana", "GOSPIĆ", "Galovac", "Galovac, Zadar", "Garesnica", "Garešnica", "Garešnica ", "Gata", "Gata/Omis", "Glina", "Glina ", "Glinska Poljana, Vratečko, Slana", "Gorica (Zadar)", "Gornja Hlapa, Dobrinj, Otok Krk", "Gornje Prilišće", "Gornji Bogićevci", "Gornji Karin", "Gornji Karin(okolica Zadra)", "Gospic", "Gospić", "Gospić ", "Grabovac", "Gracac (skradin)", "Gracac -skradin ", "Grad Zagreb", "Gradec-Vrbovec", "Gračac", "Gračac, Srb", "Gračišće, Istra", "Grubine- Imotski", "Grubišno Polje", "Gundinci", "Gunja", "Gvozd", "Hemuševec", "Herrenberg", "Hrvatski Leskovac", "Hvar", "IVANKOVO", "Igrane", "Imbriovec", "Imotski", "Imotski ", "Imotski , Cista Velika", "Imotski- Runovici ", "Ist", "Ivanec", "Ivanic-Grad", "Ivanić Grad", "Ivanić-Grad", "Ivanićko-Graberje", "Ivankovo", "Ivankovo (vinkovci)", "Ivanska", "JANJINA (PELJEŠAC)", "JESENICE", "Jabukovac", "Jadranovo", "Jakovlje", "Jaruge, Slavonski Brod", "Jastrebarsko", "Jelsa", "Jelsa/ Ivan Dolac", "Jesenice", "Jesenice(Omiš) ", "Jezera", "Jezera ", "Josipdol", "Josipovac Punitovački", "KARLOVAC", "KRAVARSKO-NOVO BRDO", "KUTINA", "Kali", "Kalinovac", "Kanfanar", "Kapelscak", "Karin Gornji", "Karlobag", "Karlovac", "Kastav", "Kastel Luksic", "Kastela", "Kaštel Gomilica", "Kaštel Kambelovac ", "Kaštel Lukšić", "Kaštel Stari", "Kaštel Štafilić", "Kaštel Štafilić ", "Kaštela (Lukšić)", "Kistanje", "Klana", "Klek", "Klinca sela", "Klis", "Klostar Ivanic", "Kloštar Ivanić", "Knezevi Vinogradi", "Konavle", "Kopacevo", "Kopačevo", "Koprivnica", "Koprivnica ", "Koprivnica - Drnje", "Koprivnički Bregi", "Korenica", "Korivnica", "Korčula", "Korčula - Brna", "Košutarica ,44324 Jasenovac", "Kraljevec na Sutli", "Kraljevica", "Kranj", "Krapina", "Krapinske Toplice", "Krašić", "Krivi Put ", "Križ", "Križevci", "Križevci ", "Krk", "Krk, Punat", "Krnjak", "Kršan", "Kukljica", "Kupljenovo (uz Zaprešić)", "Kutina", "LASLOVO/OSIJEK", "Ladimirevci", "Ladimirevci, Valpovo", "Lasinja", "Lasinja ", "Ledenice, Novi Vinodolski", "Legrad", "Lekenik", "Linardići, Krk", "Lipa", "Lipik", "Lipovac", "Lisac,Ston", "Ljubešćica", "Lokvičići ", "Lovinac", "Lovran", "Ludbreg", "Lukač", "Luzani", "Lučko", "Lužani", "MAKARSKA", "MATULJI", "MOSTAR", "MOŠĆENIČKA Draga ", "MURTER", "Majske Poljane", "Makarska", "Makarska ", "Mala Ludina", "Mala Subotica", "Malinska", "Mandre", "Mandre, otok Pag", "Marcana(Pula) ", "Marija Bistrica", "Marija Gorica", "Marina", "Marina-Trogir", "Markovac Našički", "Martinšćica", "Marusevec", "Marčana Kavran", "Maslenica", "Matulji", "Matulji / Opatija", "Matulji-Rukavac", "Medulin", "Medulin ", "Medulin/Loborika", "Metković", "Međugorje", "Mijesto Brckovljani", "Milna! Otok Brac", "Mimice", "Mirkovci", "Misenicka draga ", "Molunat", "Mostar", "Mošćenička Draga", "Mraclin", "Mrzović", "Mursko Sredisce", "Mursko Središče", "Murter ", "N. Zagreb", "NOVI OGLASI NA OVU STRANICU", "Nard", "Našice-Lađanska", "Nedelišće", "Neorić-Sutina (Split) ", "Neum", "Nin", "Nin, Žerava", "Njivice", "Nova Gradiska", "Nova Gradiška", "Nova Gradiška ", "Nova Kapela", "Novalja", "Novi Vinodolski", "Novi Vinodolski ", "Novi Vinodolski - Bribir", "Novigrad", "Novigrad (Zadarski)", "Novigrad Podravski", "Novigrad Podravski ", "Novigrad, Zadar", "Novigrad/Paljuv", "Novo Virje", "Novo Čiče", "Novoselec", "Novoselec ", "Novska", "Novska ", "OSIJEK", "OSIJEK ", "Obedišće", "Obrovac", "Obrovac ", "Ogulin", "Ogulin ", "Oklaj", "Okolica Sinja ", "Okolica Splita", "Okrug Gornji", "Okrug Gornji, Ciovo", "Okučani", "Omis", "Omiš", "Omiš,Podašpilje", "Opatija", "Opatija ", "Oprtalj", "Orahovica", "Orebic", "Orehovica", "Oroslavje", "Osijek", "Osijek ", "Osijek(Kopačevo)", "Osjek", "Otok Hvar", "Otok Krk", "Otok Pasman kod Zadra", "Otok Pasman kraj Zadra", "Otok Rab", "Otok Ugljan", "Otok Vir", "Otok vir ", "Otočac", "Otočac ", "Ozalj", "PAKOSTANE", "PETRINJA", "PIROVAC", "PITOMAČA", "PREMANTURA ", "PRIMOŠTEN", "Pag", "Pakostane", "Pakoštane", "Pakoštane ", "Pakrac", "Pakrac ", "Palešnik", "Palešnik (Garešnica) ", "Palešnik - Garešnica", "Petrijanec", "Petrinja", "Petrinja ", "Petrinja i okolica", "Petrinja i okolna mjesta", "Petrovci kraj Vukovara", "Pirovac", "Pisak", "Pisarovina", "Pitomača", "Pleternica", "Pleternica ", "Plitvice", "Plitvicka jezera", "Plitvička Jezera", "Ploce", "Ploče", "Pobri", "Podaca", "Podaca - Makarska rivijera ", "Podgora", "Podgora ", "Podstrana", "Podstrana , Split", "Podstrana Split", "Podstrana(kraj Splita)", "Polača", "Poličnik", "Poljica-Marina", "Popovac", "Popovača", "Porec", "Poreč", "Poreč ", "Posedarje", "Posedarje ", "Posedarje - Islam Latinski", "Posedarje Slivnica", "Povlja otok Brac", "Povljana", "Povljana(Pag)", "Požega", "Požega ", "Pregrada", "Pregrada ", "Preko,poljana", "Prelog", "Prezid", "Prezid, Gorski kotar", "Primošten", "Primošten ", "Privlaka", "Promajna ", "Pula", "Pula ", "Punat", "Puska", "Rab", "Radenci", "Rajevo Selo", "Rajić", "Rakitovica", "Rakovica", "Račišće 109", "Raša", "Ražanj-Rogoznica", "Retkovci", "Rešetari", "Rijeka", "Rijeka ", "Rogoznica", "Rogoznica ", "Rogoznica /Šibenik ", "Rovanjska", "Rovinj", "Rovinj ", "Rovinj j(Rovigno)", "Rovišće", "Rukavac-Matulji", "SAVUDRIJA", "SPLITSKA OTOK BRAČ", "Saborsko", "Samobor", "Sarvaš", "Seget Vranjica", "Seget Vranjica ", "Sela Sisak", "Selce", "Seline,Starigrad-Paklenica", "Selo Jezero Posavsko", "Senj", "Sesvete", "Sesvete ", "Severin na Kupi", "Sibenik", "Sinj", "Sirač ", "Sisak", "Sisak ", "Sisak (Caprag) ", "Siče", "Skopje ", "Skradin", "Skradin ", "Slana, Gkinska Poljana", "Slatina", "Slatina ", "Slatina, Gornji Miholjac", "Slavonski Brod", "Slavonski Brod ", "Slavonski Brod (Donja Bebrina)", "Slavonski Brod/Gornja Vrba", "Slavonski Kobaš ", "Slavonski brod (Podcrkavlje)", "Slovenija, adlešiči", "Slunj", "Slunj ", "Solin", "Soline", "Soljani, Vukovarsko-Srijemska", "Spisic Bukovica (Virovitica) ", "Spl", "Split", "Split ", "Split/Grude", "Split/Split", "Sreser", "Srijane", "Srima ", "Srima Vodice", "St. Toplice", "Stara Gradiska", "Stara Gradiška ", "Stara Novalja", "Stari Jankovci", "Stari Jankovci/Vinkovci", "Stari Perkovci", "Starigrad Dol", "Starigrad Paklenica", "Starigrad,Paklenica", "Starigrad-Paklenica", "Staro Petrovo Selo", "Staro Petrovo Selo, Ljudevita Posavskog 1", "Staro petrovo selo", "Stobrec", "Strizivojna", "Stupnik", "Suhopolje ", "Sukošan", "Sukošan (Zadar)", "Sukošan-Zadar", "Sumartin", "Supetar", "Supetar, Brac", "Sutivan, Brač", "Sveti Filip i Jakov", "Sveti Ivan Zelina, Beloslavec", "Sveti Martin na Muri", "Sveti Petar na Moru", "Sveti Petar na moru", "Sveti križ začretje", "Sveti martin na muri", "Svibje", "Svirče Hvar", "TROGIR", "Terezino Polje", "Tisno", "Tkon", "Tomislavgrad", "Topusko", "Tordinci", "Tribalj", "Tribunj", "Trilj", "Trogir", "Trosnj", "Trpanj", "Trpanj Pelješac", "Tršće,Čabar Gorski Kotar", "Tucepi", "Tuheljske Toplice", "Tučepi", "UMAG", "Udbina", "Umag", "Umag ", "VIR", "VRSI", "Vabriga", "Valpovo ", "Varadin", "Varazdin", "Varazdin ", "Varazdinu ", "Varaždin", "Varaždin ", "Varaždin, Radovec", "Varaždinske Toplice", "Varaždinske Toplice ", "Vela Luka", "Veli Losinj", "Velika", "Velika Gorica", "Velika Gorica ", "Velika Gorica (Kobilić) ", "Velika Graduša", "Velika Kopanica", "Velika Kopanica ", "Velika gorica ", "Veliki Grđevac / Bjelovarsko-bilogorska", "Veliko Trojstvo", "Verunić", "Vinišće", "Vinjerac ", "Vinkovci", "Vinkovci ", "Vinkovci (Sopot) ", "Vinogradci", "Vir", "Vir ", "Virje", "Virovitica", "Viskovo Rijeka ", "Visoko ", "Vivodina, Ozalj", "Viškovo", "Vocin ", "Vodice", "Vodice ", "Vodice Srima", "Vodice, ČISTA VELIKA", "Vodnjan", "Vodnjan(Dignano)", "Vodnjan~Peroj", "Vozilici", "Vođinci", "Vođinci ", "Vranjic", "Vratišinec-Međimurje", "Vrbanja(Županja)", "Vrbanja,Hrvatska", "Vrbnik", "Vrbnik ", "Vrbovec", "Vrgorac", "Vrlika, Garjak", "Vrpolje ", "Vrsar ", "Vrsi", "Vukovar", "Vukovar (Bogdanovci)", "Vukovar - Bogdanovci", "Vukovar, Ilača", "Waterford", "ZADAR", "ZADAR ", "ZAGREB", "ZAGREB( SESVETE)", "Zaagreb", "Zabok", "Zadar", "Zadar ", "Zadar (Briševo)", "Zadar (Crna Punta) ", "Zadar Preko- Poljana", "Zadar( Sukošan )", "Zadar, Sv.Filip i Jakov", "Zadarska zupanija", "Zagreb", "Zagreb ", "Zagreb - Dubrava", "Zagreb - Vrbovec", "Zagreb / Lucko", "Zagreb Kašina", "Zagreb Perjavica", "Zagreb Sesvete", "Zagreb,Sesvete", "Zagreb-Novi Zagreb", "Zagrebacka zupanija Varazdinska Krapinsko zagorska ", "Zaprešić", "Zašrešić", "Zelina", "Zemunik Donji (Zadar)", "Zg", "Zlatar Bistrica", "Zubovići ", "bregana", "dalj", "dubrovnik", "gg", "komin neretva", "mali Lošinj", "mali losinj", "marčana", "okučani", "primosten", "seget", "sibenik", "soljani", "split", "tuzla", "zadar", "zagreb", "Čakovec", "Čavle", "Čavle ", "Čavle-Cernik 13(okolica Rijeke)", "Čazma", "Čačinci", "Čačinci ", "Čađavački Lug / Slatina", "Čepin", "Čepinski Martinci", "Čilipi", "Čiovo", "Čiovo - Trogir", "Čiovo, Zagreb", "Črnkovci ", "Črnkovec", "čakovec", "Đakovo", "Đelekovec ", "Đurđevac", "Đurđevac ", "ŠIBENIK", "ŠIBENik", "ŠMRIKA", "Šandrovac", "Šarengrad", "Šegotići, Marčana, Pula ", "Šestanovac", "Šibenik", "Šibenik ", "Šibenik, Rogoznica", "Šilo (Krk)", "Širitovci/Drniš", "Šišinec", "Škabrnja", "Škrip (otok Brač)", "Škropeti", "Šmrika", "Šolta", "Štitar", "Štrigova", "šibenik", "Ždrelac (Otok Pašman)", "Živogošće", "Županja"]
```

vfonic commented:
I hope I've addressed all the comments now. :) I hoped we'll merge this temporary issue mitigation three days ago.

I think we're reaching the point where we either merge this or just drop it altogether and wait for the better/more complete solution. What do you think?

I'm happy to make further minor changes, but keep in mind that testing this, even with the partial automated test I've written, is a time-consuming and error-prone task.

Here's the test I used to speed up testing of this functionality:
https://github.com/vlado/earthquake-croatia/pull/41/commits/30d7fcf718ae276b9f0b9ba6972362581c88079e#diff-03ecf6950cb3efe648f741b8578890b9265814b4196c05af83f76dfda88058fa

You can find "old" => "new" cities JSON array in the same commit.

/cc @vlado @shime

mmorava commented:
@vlado can you please run my migration from https://github.com/vlado/earthquake-croatia/pull/102 on a production app after normalization? I believe there will be issues that needs to be resolved I wonder how many of them

vfonic commented:
@mmorava isn't it that the migration will fail on the first city it cannot find by name and stop execution and the transaction will be rolled back?

vfonic commented:
@mmorava you can try to create a new branch (from your PR branch) and rebase it on top of this and see what happens.

vfonic commented:
@mmorava here's the list of all the cities in the live app as of 11 h ago:
https://github.com/vlado/earthquake-croatia/pull/41#issuecomment-753589906t

vlado commented on db/migrate/20210103163430_add_index_to_kind_and_service_on_ads.rb:
```diff
@@ -0,0 +1,6 @@
+class AddIndexToKindAndServiceOnAds < ActiveRecord::Migration[6.1]
+  def change
+    add_index :ads, :kind
+    add_index :ads, :service
```
This one was renamed (again) to `category` in master https://github.com/vlado/earthquake-croatia/pull/96. Sorry about that.

vfonic replied:
:) No worries. I was more surprised to see repurposing `kind` into a completely different concept.

vlado commented on db/migrate/20210103163430_add_index_to_kind_and_service_on_ads.rb:
```diff
@@ -0,0 +1,6 @@
+class AddIndexToKindAndServiceOnAds < ActiveRecord::Migration[6.1]
+  def change
+    add_index :ads, :kind
```
Should be already there from previous migration.

vlado reviewed (commented):
I think this is valuable and would like to merge this cause I think they are quite ready (see my comments) and brings improvements.

vlado commented:
I suggest we test this by copying production db to review app. Can you ping me on Slack when you are ready and we do this together?

vfonic commented:
@vlado I believe I fixed everything now. Rebased and force-pushed.

Please go over [cities_mapping.yml](https://github.com/vfonic/earthquake-croatia/blob/normalize-db-data/lib/tasks/cities_mapping.yml).

I was "normalizing" city names to the best of my knowledge. Please note "Sesvete" => "Zagreb" and I also tried to rename to the smallest possible village name I could figure out from the city name. I think that might not be the best solution with the recent developments... Feel free to take over and push any new commits to this PR/branch. I gave you write access on my fork.
